Feature/drive cancellation and enable 10th UHF channel

antsr opened the pull request:
Changes proposed in this pull request:
- Add drive crosstalk cancellation tuneup routines
- Enable 10th integration channel of the UHF 

Could you review @chellings ?

antsr committed:
initial implementation of drive cancellation pulse and tuneup measurement
untested

antsr committed:
add analysis for drive xtalk cancellation

antsr committed:
bugfix cancellation drive
enable 10th integration channel on the UHFQA

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+            n_per_state=n_cal_points_per_state)
+        operation_dict = dev.get_operation_dict()
+
+        seq, sweep_vals = mqs.drive_cancellation_seq(
```
please add comment:
```
# We get sweep_vals for only one dimension since drive_cancellation_seq
# turns 2D sweep points into 1D-SegmentHardSweep.
# FIXME: in the future, this should rather be implemented via
# sequence.compress_2D_sweep
```

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+        """
+        Sweep pulse cancellation parameters and measure Ramsey on cancelled qubits.
+        The sweep point keys should be of the form `qb.param`, where `qb` is the
+        name of the qubit the cancellation if for and `param` is a parameter in
```
qubit the cancellation **is** for

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+        n_cal_points_per_state=2, cal_states='auto', prep_params=None,
+        exp_metadata=None, label=None, upload=True, analyze=True):
+        """
+        Sweep pulse cancellation parameters and measure Ramsey on cancelled qubits.
```
cancelled qubits --> qubits the cancellation is for

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+        you could configure the sweep points as `SweepPoints('qb1.amplitude',
+        np.linspace(0, 1, 21))`.
+
+        The second sweep dimension of sweep_points must be called 'phases'.
```
sweep_points have to be 1dim and phases can be passed a parameter here

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+        np.linspace(0, 1, 21))`.
+
+        The second sweep dimension of sweep_points must be called 'phases'.
+        """
```
please explain the parameters in the docstring

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
 
+def measure_drive_cancellation(
+        dev, driven_qubit, ramsey_qubits, sweep_points,
+        phases=np.linspace(0, 2*np.pi, 3, endpoint=False), n=1, pulse='X180',
```
maybe a self-explanatory parameter name can be used instead of "n"

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+
+        det_func = get_multiplexed_readout_detector_functions(
+            ramsey_qubits,
+            nr_averages=max([qb.acq_averages() for qb in ramsey_qubits]))\
```
Isn't this choice for nr_averages the default in get_multiplexed_readout_detector_functions anyways? (no need to pass it)

chellings commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1788,6 +1788,49 @@ def measurement_induced_dephasing_seq(
+    np.linspace(0, 1, 21))`.
+
+    The second sweep dimension of sweep_points must be called 'phase'.
+    """
```
please explain params in docstring, at least those that are specific to this measurements. (maybe then write: all other params are passed to interleaved_pulse_list_list_equatorial_seq)

chellings commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1788,6 +1788,49 @@ def measurement_induced_dephasing_seq(
+def drive_cancellation_seq(
+        driven_qubit_name, ramsey_qubit_names, operation_dict,
+        sweep_points, pulse='X180', n=1, pihalf_spacing=None, prep_params=None,
+        cal_points=None, upload=True, sequence_name='drive_cancellation_seq'):
```
It might be cleaner to have a single parameter drive_op_code instead of two parameters driven_qubit_name and pulse

chellings commented on pycqed/measurement/waveform_control/pulse_library.py:
```diff
@@ -113,6 +113,112 @@ def hashables(self, tstart, channel):
+                iq_idx = [i, q].index(channel)
+                cpars = p
+                break
+        assert iq_idx != -1
```
why not return np.zeros_like(tvals) if this happens instead of raising an AssertionError? (at least this is the behavior of SSB_DRAG_pulse.chan_wf)

chellings commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -46,7 +46,7 @@ class UHFQCPulsar:
     
     _uhf_sequence_string_template = (
-        "const WINT_EN   = 0x01ff0000;\n"
+        "const WINT_EN   = 0x03ff0000;\n"
```
No idea about this, but I trust you 😉

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -1843,6 +1843,101 @@ def measure_measurement_induced_dephasing(qb_dephased, qb_targeted, phases, amps
+        len_sweep = len(list(sweep_points[0].values())[0][0])
+        sweep_points_for_analysis[0]['phase'] = \
+            (np.tile(phases, len_sweep), 'rad', 'Ramsey phase')
+        meas_obj_sweep_points_map = {qbn: ['phase'] for qbn in ramsey_qubit_names}
```
@stephlazar might have a comment on this when reviewing the analysis part.

stephlazar replied:
Yes, this won't work with the fix that I added last week. The concept for SweepPoints is that they should always describe the experiment conceptually. So if your experiment is a 2D sweep conceptually, then the SweepPoints should have 2 dimensions. And the meas_obj_sweep_points_map for each measured object should have at least one key from each dimension. I will try to fix things myself and will get in touch with you @antsr if I can't figure it out.

stephlazar replied:
Fixed now.

chellings reviewed (changes requested):
Thank you for this useful feature which works already nicely on XLD. My comments mainly refer to docstrings and attribute names. Apart from this, everything looks fine to me in the measurement, sequence creation, pulse library, and in the general UHF-related changes.

After a quick chat with @stephlazar, she volunteered to review the analysis part, so I have not looked at the analysis in the tda module. I have checked the changes in all other modules in detail.

stephlazar commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -3875,6 +3875,204 @@ def prepare_plots(self):
+                continue
+            qb, param = k.split('.')
+            pdd['qb_sweep_points'][qb] = v[0][::len(pdd['ramsey_phases'])]
+            pdd['qb_sweep_param'][qb] = (param, v[1], v[2])
```
I will try to fix the above lines that handle SweepPoints after I fix their creation in the measure_function. Ideally, these lines can be completely removed in that case.

stephlazar replied:
Fixed now.

Commit:
Merge branch 'qudev_master' into feature/drive_cancellation

Commit:
Fixes the use of sweep points in measure_drive_cancellation and DriveCrosstalkCancellationAnalysis + improves get_meas_obj_sweep_points_map.

Removes sweep_points_for_analysis from measure_drive_cancellation, and calls analysis with TwoD == True and the two-dimensional SweepPoints.
 This means this measurement is now detected as a hybrid measurement in BaseDataAnalysis.add_measured_data().
Improves SweepPoints.get_meas_obj_sweep_points_map() to first try to look for mobj_names in the sweep parameter names, and to append
 the same sweep param name to all qubits if a dimension dictionary has only one key.

stephlazar commented:
@antsr @chellings I have now fixed the measure function + analysis to work with the code I have added in BaseDataAnalysis for hybrid measurements. I tested that the measurement runs in a virtual setup, and I successfully reanalyzed some of the data files Ants took. 

**So this pull request is approved from my side.**

antsr committed:
update docstrings

antsr commented:
I now addressed @chellings comments. But the measurement and analysis should be tested again before merging.

antsr committed:
add

antsr committed:
fixes to drive cancellation measurement

chellings committed:
Merge branch 'qudev_master' into feature/drive_cancellation

antsr committed:
Merge branch 'qudev_master' into feature/drive_cancellation

antsr commented:
The code was tested on the XLD S17 branch on 10. July and the bugfixes are now included in this PR.
